cpp/src/gandiva/decimal_ir.cc

```diff
@@ -307,6 +307,52 @@ Status DecimalIR::BuildAdd() {
   return Status::OK();
 }
 
+Status DecimalIR::BuildSubtract() {
+  // Create fn prototype :
+  // int128_t
+  // subtract_decimal128_decimal128(int128_t x_value, int32_t x_precision, int32_t
+  // x_scale,
+  //                           int128_t y_value, int32_t y_precision, int32_t y_scale
+  //                           int32_t out_precision, int32_t out_scale)
+  auto i32 = types()->i32_type();
+  auto i128 = types()->i128_type();
+  auto function = BuildFunction("subtract_decimal128_decimal128", i128,
+                                {
+                                    {"x_value", i128},
+                                    {"x_precision", i32},
+                                    {"x_scale", i32},
+                                    {"y_value", i128},
+                                    {"y_precision", i32},
+                                    {"y_scale", i32},
+                                    {"out_precision", i32},
+                                    {"out_scale", i32},
+                                });
+
+  auto entry = llvm::BasicBlock::Create(*context(), "entry", function);
+  ir_builder()->SetInsertPoint(entry);
+
+  // reuse add function after negating y_value. i.e
+  //   add(x_value, x_precision, x_scale, -y_value, y_precision, y_scale,
+  //       out_precision, out_scale)
+  std::vector<llvm::Value*> args;
+  int i = 0;
+  for (auto& in_arg : function->args()) {
+    if (i == 3) {
+      auto y_neg_value = ir_builder()->CreateNeg(&in_arg);
```

Comment:
I was wondering if this could overflow, but the decimal type is limited to 38 decimal digits, which is too small for `2**127`.

As a side note, it seems we don't validate inputs to the `Decimal128` type and happily accept larger numbers:
```python
>>> ty = pa.decimal128(39, 0)                                                                            
>>> arr = pa.array([(2**127) - 1], type=ty)                                                              
>>> arr                                                                                                  
<pyarrow.lib.Decimal128Array object at 0x7f9b89444e08>
[
  170141183460469231731687303715884105727
]
>>> arr = pa.array([(2**127)], type=ty)                                                                  
>>> arr                                                                                                  
<pyarrow.lib.Decimal128Array object at 0x7f9b89444138>
[
  -170141183460469231731687303715884105728
]
>>> arr = pa.array([-(2**127)], type=ty)                                                                 
>>> arr                                                                                                  
<pyarrow.lib.Decimal128Array object at 0x7f9b89432688>
[
  -170141183460469231731687303715884105728
]
```

Reply:
I'll create a jira to ensure gandiva fails when the type has a precision/scale is > 38.

For the actual values itself, I can add a IsValid() api to BasicDecimal (that checks against a min/max value). but, checking this each time will be a perf overhead. maybe, we should have a conf variable to do overflow checks.

Reply:
Intuitively, I'd say people who use decimals value correctness. Though it's clear our C++ API currently doesn't check...

Reply:
I've created ARROW-4569 and ARROW-4570 to follow up on these.